clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -2743,9 +2755,22 @@ int main(int Argc, char **Argv) {
       if (Error Err = sycl::mergeSYCLBIN(*FilesOrErr, Args))
         reportError(std::move(Err));
     } else {
-      // Run the host linking job with the rendered arguments.
-      if (Error Err = runLinker(*FilesOrErr, Args))
-        reportError(std::move(Err));
+      if (Args.hasArg(OPT_sycl_device_link)) {
+        // Skip host linker if --sycl-device-link option is set.
+        // Just copy the output of device linking and wrapping action.
+        if (FilesOrErr->size() == 1) {
+          if (Error Err =
+                  sycl::copyFileToFinalExecutable((*FilesOrErr)[0], Args))
+            reportError(std::move(Err));
+        } else {
+          reportError(createStringError(
+              "Expect single output from the device linker."));
+        }
+      } else {
+        // Run the host linking job with the rendered arguments.
+        if (Error Err = runLinker(*FilesOrErr, Args))
+          reportError(std::move(Err));
+      }
```

Comment:
nit: I think this section would be easier to read if we remove some indentation.

```c++
if (OutputSYCLBIN) {
  if (Error Err = sycl::mergeSYCLBIN(*FilesOrErr, Args))
    reportError(std::move(Err));
} else if (Args.hasArg(OPT_sycl_device_link)) {
  if (FilesOrErr->size() != 1)
    reportError(createStringError("Expect single output from the device linker."));
  if (Error Err = sycl::copyFileToFinalExecutable((*FilesOrErr)[0], Args))
    reportError(std::move(Err));
} else {
  if (Error Err = runLinker(*FilesOrErr, Args))
    reportError(std::move(Err));
}
```

Reply:
Spoke to Arvind offline about this and I think the best course of action for now is to diagnose the case where SYCLBIN has been requested at the same time as the link flag is passed. Eventually we want it to work, but only for cases where the inputs are SYCLBIN files, as otherwise we would have some awkward cases where we would be dropping host code. I plan to work on SYCLBIN linking support in the near future.